### PR TITLE
[FinalCutProXml15] - Update

### PR DESCRIPTION
**Pull request by ivandrofly**

### CODE REVIEW

File: libse/SubtitleFormats/FinalCutProXml15.cs [hunk 1]
```diff
@@ -469,39 +475,22 @@ public override void LoadSubtitle(Subtitle subtitle, List<string> lines, string
...
+            var arr = trimmedDuration.Split('/');
+            if (arr.Length == 2)
             {
-                var dur = duration.Value;
-                if (dur.EndsWith("ms"))
-                {
-                    var arr = duration.Value.TrimEnd('s').TrimEnd('m').Split('/');
-                    if (arr.Length == 2)
-                    {
-                        return TimeCode.FromSeconds((long.Parse(arr[0]) * 1000.0) / (double.Parse(arr[1]) * 1000.0));
```
Comment (ivandrofly) [inline libse/SubtitleFormats/FinalCutProXml15.cs]: 1000.0 cancels out

File: libse/SubtitleFormats/FinalCutProXml15.cs [hunk 2]
```diff
@@ -469,39 +475,22 @@ public override void LoadSubtitle(Subtitle subtitle, List<string> lines, string
...
-                    }
-                    if (arr.Length == 1)
-                    {
-                        return TimeCode.FromSeconds(float.Parse(arr[0]) * 1000.0);
```
Comment (ivandrofly) [inline libse/SubtitleFormats/FinalCutProXml15.cs]: The method returns new TimeCode from seconds why converting it to milliseconds?

File: libse/SubtitleFormats/FinalCutProXml15.cs [hunk 3]
```diff
@@ -193,35 +194,41 @@ public override string ToText(Subtitle subtitle, string title)
...
                 {
-                    var ch = text[i];
-                    if (ch == '<' && text.Substring(i).StartsWith("<i>", StringComparison.OrdinalIgnoreCase))
+                    char ch = text[i];
+                    if (ch != '<')
+                    {
+                        sb.Append(ch);
+                        continue;
+                    }
+                    string tempText = text.Substring(i);
```
Comment (ivandrofly) [inline libse/SubtitleFormats/FinalCutProXml15.cs]: Better than keep constructing new string once a statement fails...

File: libse/SubtitleFormats/FinalCutProXml15.cs [hunk 4]
```diff
@@ -180,10 +180,11 @@ public override string ToText(Subtitle subtitle, string title)
                 ;
             XmlNode videoNode = xml.DocumentElement.SelectSingleNode("//project/sequence/spine");
             int number = 1;
+            var trimmedTitle = new StringBuilder();
             foreach (Paragraph p in subtitle.Paragraphs)
             {
                 XmlNode video = xml.CreateElement("video");
-                var trimmedTitle = new StringBuilder();
+                trimmedTitle.Clear();
```
Comment (ivandrofly) [inline libse/SubtitleFormats/FinalCutProXml15.cs]: More optimized

File: libse/SubtitleFormats/FinalCutProXml15.cs [hunk 5]
```diff
@@ -469,39 +475,22 @@ public override void LoadSubtitle(Subtitle subtitle, List<string> lines, string
             }
         }
 
-        private static TimeCode DecodeTime(XmlAttribute duration)
+        private static TimeCode DecodeTime(XmlAttribute xduration)
```
Comment (niksedk) [inline libse/SubtitleFormats/FinalCutProXml15.cs]: Ups, this is not right - "ms" is milliseconds and "s" is seconds - also "duration" or similar is better than "xduration"...

Comment (ivandrofly) [reply libse/SubtitleFormats/FinalCutProXml15.cs]:
```c#
if (duration != null)
{
	var dur = duration.Value;
	if (dur.EndsWith("ms"))
	{
		var arr = duration.Value.TrimEnd('s').TrimEnd('m').Split('/');
		if (arr.Length == 2)
		{
			return TimeCode.FromSeconds((long.Parse(arr[0]) * 1000.0) / (double.Parse(arr[1]) * 1000.0));
		}
		if (arr.Length == 1)
		{
			return TimeCode.FromSeconds(float.Parse(arr[0]) * 1000.0);
		}
	}
```

Are we supposed to divide by **1000** to get to seconds instead?
e.g: return TimeCode.FromSeconds((long.Parse(arr[0]) **/** 1000.0) / (double.Parse(arr[1]) **/** 1000.0));

/cc @niksedk

**Comment by ivandrofly** [comment]
UPDATED /CC @niksedk

File: libse/SubtitleFormats/FinalCutProXml15.cs [hunk 6]
```diff
@@ -438,24 +444,24 @@ public override void LoadSubtitle(Subtitle subtitle, List<string> lines, string
...
-                            bool add = true;
+                            p.Text = text;
                             if (subtitle.Paragraphs.Count > 0)
                             {
                                 var prev = subtitle.Paragraphs[subtitle.Paragraphs.Count - 1];
-                                if (prev.Text == p.Text && prev.StartTime.TotalMilliseconds == p.StartTime.TotalMilliseconds)
-                                    add = false;
+                                if (prev.Text.Equals(p.Text, StringComparison.Ordinal) && prev.StartTime.TotalMilliseconds == p.StartTime.TotalMilliseconds)
+                                {
+                                    continue; // ignore duplicated paragraphs
```
Comment (niksedk) [inline libse/SubtitleFormats/FinalCutProXml15.cs]: Why have you added this?

Comment (ivandrofly) [reply libse/SubtitleFormats/FinalCutProXml15.cs]: We don't need variable **add** to check if paragraph should be added

```c#
var p = new Paragraph();
p.StartTime = DecodeTime(node.ParentNode.Attributes["offset"]);
p.EndTime.TotalMilliseconds = p.StartTime.TotalMilliseconds + DecodeTime(node.ParentNode.Attributes["duration"]).TotalMilliseconds;
p.Text = text;
if (subtitle.Paragraphs.Count > 0)
{
	var prev = subtitle.Paragraphs[subtitle.Paragraphs.Count - 1];
	if (prev.Text.Equals(p.Text, StringComparison.Ordinal) && prev.StartTime.TotalMilliseconds == p.StartTime.TotalMilliseconds)
	{
		continue; // ignore duplicated paragraphs
	}
}
subtitle.Paragraphs.Add(p);
```

**Pull request closed without merging**